### ISSUE TITLE
Migration from event-based system

### ISSUE THREAD

**Issue by ilyazub** [rackt/redux#860]
Few days ago I've seen talk from the StrangeLoop conference about [Redux + React usage within FF DevTools](https://youtu.be/qUlRpybs7_c). Before that I thought that migration from system, based on the state mutations, to Redux is almost impossible.

In project I'm currently working on we're using Backbone.js. System should be refactored regardless of migration to Redux. I'm wondering about steps to migrate Backbone.js-like (shared mutable state, etc.) to unidirectional system, besides [rewriting model layer](https://rackt.github.io/redux/docs/recipes/MigratingToRedux.html#from-backbone).

I guess, the main thing I should do is to eliminate bi-directional relations between models and controller-views, and eliminate shared mutable state (which are almost the same issues in my opinion).

The one step to achieve it is to move all global (`Backbone.on('eventName')`) events to the single object and rely on this object instead of listening to events (something like action creators, [EventAggregator](http://martinfowler.com/eaaDev/EventAggregator.html)). I can't put into words the next steps, therefore, I can't achieve it.

So, I'm asking for help in compiling the list of steps in migration from shared mutable state to the single source of truth. Sounds strange, I know.

**Comment by somebody32** [comment]
We started the same migration a couple of months ago, and now part of our app is fully working with react+redux and coexisting without any problems with other legacy backbone/marionette parts in production.

The main problem was how to sync all backbone collections with redux tree. We decided that it will be easier to threat BB-collections as a single source of truth. We created `ears`, special functions that listen for collections' events and triggering action creators depending on the event type. Action creators trigger 4 types of actions: `add`, `remove`, `merge` and `reset` and serializing models just into plain JSON.
Then reducers will listen to this actions and store serialized objects in the tree using flat structure like this: 

```javascript
{
  entities: [{model1}, {model1}],
  by_id: {
    1: {model1},
    2: {model2},
  },
  ...other indexes
}
```

This allow us to store data in the way UI needs it, not how server dictates us to store it.

And the last part of the schema is `bridges`: backbone views with overridden `render` method. This allow us to migrate step-by-step and very granularly to react, spreading it like a virus. In this `render` method you just need to call `React.render` with you container wrapped in `<Provider>` from `react-redux` and you're done. Inside container you connect to redux tree normally without any hacks.

So now we can keep new and legacy parts of application in isolation and not passing any backbone hacks/wrappers into react world. All updates are coming from redux, and all UI-interactions are just functions that call BB-models, but react is not interested in any output/side-effects of this calls.

I also found that our ears/action creators/reducers for collections were very repetitive, so I extracted them into a separate lib: [backbone-redux](https://github.com/redbooth/backbone-redux/). 
Current master is outdated, but new version is around the corner (I'm currently finishing documentation for it), so if you want to give at a spin, install version `0.1.0-rc2` and read docs from this [PR/branch](https://github.com/redbooth/backbone-redux/pull/5).

I know that this comment may lack a lot of small details, so feel free to ask anything
